### PR TITLE
Testing guide

### PR DESCRIPTION
**Pull request by willmcgugan**
Adds Testing to the guide.

This covers a lot. In the future I might add a few more examples.

Still do do is the snapshot framework, which I'll leave for @darrenburns

### CODE REVIEW

File: docs/guide/testing.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+!!! tip "Learn more about testing"
+
+    We recommend [Python Testing with pytest](https://pythontest.com/pytest-book/) for a comprehensive guide to writing tests.
+
+## Do you need to write tests?
+
+The short answer is "no", you don't *need* to write tests.
+
+In practice however, it is almost always a good idea to write tests.
```
Comment (brokenshield) [inline docs/guide/testing.md:19]:
```suggestion
1. In practice however, it is almost always a good idea to write tests.
```

File: docs/guide/testing.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+Writing code that is completely bug free is virtually impossible, even for experienced developers.
```
Comment (brokenshield) [inline docs/guide/testing.md:20]:
```suggestion
1. Writing code that is completely bug free is virtually impossible, even for experienced developers.
```

File: docs/guide/testing.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+If you want to have confidence that your application will run as you intended it to, then you should write tests.
```
Comment (brokenshield) [inline docs/guide/testing.md:21]:
```suggestion
1. If you want to have confidence that your application will run as you intended it to, then you should write tests.
```

File: docs/guide/testing.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+Your test code will help you find bugs early, and alert you if you accidentally break something in the future.
```
Comment (brokenshield) [inline docs/guide/testing.md:22]:
```suggestion
1. Your test code will help you find bugs early, and alert you if you accidentally break something in the future.
```

File: docs/guide/testing.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+The `run_test()` method is an *async context manager* which returns a [`Pilot`][textual.pilot.Pilot] object.
+You can use this object to interact with the app as if you were operating it with a keyboard and mouse.
+
+Let's look at the tests for the example above:
+
+```python title="test_rgb.py"
+--8<-- "docs/examples/guide/testing/test_rgb.py"
+```
+
+1. The `run_test()` method requires that it run in a coroutine, so test must use the `async` keyword.
```
Comment (brokenshield) [inline docs/guide/testing.md:65]:
```suggestion
1. The `run_test()` method requires that it run in a coroutine, so tests must use the `async` keyword.
```

File: docs/guide/testing.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+2. This runs the app and returns a Pilot instance we can use to interact with it.
+3. Simulates pressing the ++r++ key.
+4. This checks that pressing the ++r++ key has resulted in the background color changing.
+5. Simulates clicking on the widget with an `id` of `red` (the button labelled "Red").
+
+There are two tests defined in `test_rgb.py`.
+One to test keys and one to test button clicks.
```
Comment (brokenshield) [inline docs/guide/testing.md:72]:
```suggestion
The first to test keys and the second to test button clicks.
```

File: docs/guide/testing.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+Both tests first construct an instance of the app and call `run_test()` to get a Pilot object.
```
Comment (brokenshield) [inline docs/guide/testing.md:73]:
```suggestion
Both tests first construct an instance of the app and then call `run_test()` to get a Pilot object.
```

File: docs/guide/testing.md [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+```
+
+Each string creates a single keypress.
+You can also use the name for non-printable keys (such as "enter") and the "ctrl+" modifier.
+These are the same identifiers as used for key events, which you can experiment with by running `textual keys`.
+
+## Simulating clicks
+
+You can simulate mouse clicks in a similar way with [`Pilot.click`][textual.pilot.Pilot.click].
+If you supply a css selector Textual will simulate clicking on the matching widget.
```
Comment (brokenshield) [inline docs/guide/testing.md:101]:
```suggestion
If you supply a CSS selector Textual will simulate clicking on the matching widget.
```

File: docs/guide/testing.md [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+```python
+async with app.run_test(size=(100, 50)) as pilot:
+    ...
+```
+
+## Pausing the pilot
+
+Some actions in a Textual app won't change the state immediately.
+For instance messages may take a moment to bubble from the widget that sent them.
```
Comment (brokenshield) [inline docs/guide/testing.md:158]:
```suggestion
For instance, messages may take a moment to bubble from the widget that sent them.
```

File: docs/guide/testing.md (hunk 1 above)
Comment (willmcgugan) [reply docs/guide/testing.md:19]: Not keen on a list here. Wasn't intended to be list like.

**Comment by willmcgugan** [comment]
Thanks @brokenshield

**Pull request merged**